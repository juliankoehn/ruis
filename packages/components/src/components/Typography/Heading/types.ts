export const defaultProps = {
    is: 'h2',
    testId: 'ruids-heading',
}

export type HeadingTypes = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6'

export type HeadingProps = {
    is: HeadingTypes
    testId?: string
    style?: React.CSSProperties
    className?: string
    children?: React.ReactNode
} & typeof defaultProps