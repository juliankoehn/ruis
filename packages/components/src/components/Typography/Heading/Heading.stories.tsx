import React from 'react'
import { withKnobs } from '@storybook/addon-knobs'
import { Heading } from './Heading'
import { HeadingTypes } from './types'

export default {
    title: 'Typography/Heading',
    decorators: [withKnobs]
}

const HeadingSizes: HeadingTypes[] = ['h1', 'h2', 'h3', 'h4', 'h5', 'h6']

export const Default = () => {
    return (
        <div>
            {HeadingSizes.map(size => (
                <Heading is={size}>Heading {size}</Heading>
            ))}
        </div>
        
    )
}
