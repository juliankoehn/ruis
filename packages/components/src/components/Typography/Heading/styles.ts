import { css } from '@emotion/core'
import tokens from '@ruids/tokens'
import { HeadingTypes } from './types';

const getFontStyles = (element: HeadingTypes) => {
    switch (element) {
        case 'h1': {
            return {
                size: tokens.text6Xl,
                weight: tokens.fontNormal,
                leading: tokens.leadingTight
            }
        }
        case 'h2': {
            return {
                size: tokens.text5Xl,
                weight: tokens.fontNormal,
                leading: tokens.leadingTight
            }
        }
        case 'h3': {
            return {
                size: tokens.text4Xl,
                weight: tokens.fontMedium,
                leading: tokens.leadingTight
            }
        }
        case 'h4': {
            return {
                size: tokens.text3Xl,
                weight: tokens.fontSemibold,
                leading: tokens.leadingTight
            }
        }
        case 'h5': {
            return {
                size: tokens.text2Xl,
                weight: tokens.fontBold,
                leading: tokens.leadingTight
            }
        }
        case 'h6': {
            return {
                size: tokens.textXl,
                weight: tokens.fontBold,
                leading: tokens.leadingTight
            }
        }
    }
}

export const getHeadingStyles = (spacing: string, element: HeadingTypes) => {
    const fontStyle = getFontStyles(element)
    return css`
        font-family: ${tokens.fontSans};
        color: ${tokens.colorTextDark};
        margin-top: 0;
        margin-bottom: ${tokens[spacing]};
        font-weight: ${fontStyle?.weight};
        font-size: ${fontStyle?.size};
        line-height: ${fontStyle?.leading};
    `
};