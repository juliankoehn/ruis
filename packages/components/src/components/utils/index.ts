export { canUseDOM } from './canUseDom'
export { lightenColor } from './color.lighten'
export {Â axe } from './axe'