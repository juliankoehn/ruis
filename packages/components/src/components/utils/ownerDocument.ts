export const ownerDocument = (node: Node): Document => {
  return (node && node.ownerDocument) || document;
};
