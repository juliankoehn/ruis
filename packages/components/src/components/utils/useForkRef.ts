import { useMemo, MutableRefObject } from 'react';

export type HTMLElementOrNull = HTMLElement | null;
export type CallbackRef = (node: HTMLElement | null) => any;
export type AnyRef = CallbackRef | MutableRefObject<HTMLElementOrNull>;

function setRef(ref: AnyRef, value: HTMLElementOrNull) {
  if (typeof ref === 'function') {
    ref(value);
  } else if (ref) {
    ref.current = value;
  }
}

function useForkRef(refA: AnyRef, refB: AnyRef): CallbackRef | null {
  /**
   * This will create a new function if the ref props change and are defined.
   * This means react will call the old forkRef with `null` and the new forkRef
   * with the ref. Cleanup naturally emerges from this behavior
   */
  return useMemo(() => {
    if (refA == null && refB == null) {
      return null;
    }
    return (refValue: HTMLElementOrNull) => {
      setRef(refA, refValue);
      setRef(refB, refValue);
    };
  }, [refA, refB]);
}

export { useForkRef };
