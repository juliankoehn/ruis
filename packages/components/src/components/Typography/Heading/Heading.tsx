import React from 'react'
import { HeadingProps, defaultProps } from './types'
import { TypographyContext } from '../Typography'
import { getHeadingStyles } from './styles'

export class Heading extends React.Component<HeadingProps> {
    static defaultProps = defaultProps
    render() {
        const {
            className,
            children,
            is,
            testId,
            ...otherProps
        } = this.props
        const Element = is
        return (
            <TypographyContext.Consumer>
                {value => (
                    <Element
                        className={className}
                        data-test-id={testId}
                        {...otherProps}
                        css={getHeadingStyles(
                            value.heading.spacing,
                            is
                        )}
                    >
                        {children}
                    </Element>
                )}
            </TypographyContext.Consumer>
        )
    }
}