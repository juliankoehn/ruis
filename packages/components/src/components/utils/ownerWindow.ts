import { ownerDocument } from './ownerDocument';

export const ownerWindow = (node: Node): Window => {
  const doc = ownerDocument(node);
  return doc.defaultView || window;
};
