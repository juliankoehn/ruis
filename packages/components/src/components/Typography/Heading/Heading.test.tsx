import React from 'react'
import renderer from 'react-test-renderer'
import { Heading } from './Heading'
import { render } from '@testing-library/react'
import {Â axe } from '../../utils'

test('renders the component', () => {
    const output = renderer
        .create(<Heading>Heading</Heading>)
        .toJSON()

    expect(output).toMatchSnapshot();
})

it('renders the component with an additional class name', () => {
    const output = renderer
        .create(<Heading className="my-extra-class">Heading</Heading>)
        .toJSON()

    expect(output).toMatchSnapshot();
});

it('renders the component h3', () => {
    const output = renderer
        .create(<Heading is="h3">Heading</Heading>)
        .toJSON()

    expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
    const { container } = render(<Heading>Heading</Heading>)

    expect(await axe(container)).toHaveNoViolations()
});