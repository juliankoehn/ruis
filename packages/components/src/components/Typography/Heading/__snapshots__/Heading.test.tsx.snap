// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<h2
  className="css-41y8r1"
  data-test-id="ruids-heading"
>
  Heading
</h2>
`;

exports[`renders the component h3 1`] = `
<h3
  className="css-f4lcri"
  data-test-id="ruids-heading"
>
  Heading
</h3>
`;

exports[`renders the component with an additional class name 1`] = `
<h2
  className="my-extra-class css-41y8r1"
  data-test-id="ruids-heading"
>
  Heading
</h2>
`;
